File: 1164.cpp
#include <iostream>
#include <deque>
#include <vector>
#include <algorithm>
#include <cstdint>

int main(){
    int n;
    std::cin >> n;

    std::vector<uint64_t> rooms (2*n);
    uint64_t n_dim = 2 * (uint64_t) n;
    uint64_t i_dim = n;
    for (int i = 0; i < n; i++){
        unsigned int a, b;
        std::cin >> a >> b;
        rooms[2 * i] = n_dim * a + i;
        rooms[2 * i + 1] = n_dim * b + i_dim + i;
    }

    std::sort(rooms.begin(), rooms.end());

    // for (auto val : rooms) std::cout << "(" << val[0] << ", " << val[1] << ", " << val[2] << ") ";
    // std::cout << std::endl;

    std::vector<int> assigned (n);
    int room_n = 0;
    std::deque<int> available{};

    for (auto room : rooms){
        uint64_t val = room % n_dim;
        uint64_t index = val % i_dim;
        uint64_t is_departure = val / i_dim;

        if (is_departure){
            available.push_back(assigned[index]);
        } else {
            if (available.size() == 0){
                room_n ++;
                available.push_back(room_n);
            }
            assigned[index] = available.back();
            available.pop_back();
        }
    }

    std::cout << room_n << '\n';
    for (auto val : assigned) std::cout << val << " ";
    std::cout << std::endl;
}

File: 1164.py
from collections import deque

n = int(input())
s = []
for i in range(n):
    a, b = map(int, input().split())
    s.append((i, a, -1))
    s.append((i, b, 1))

s.sort(key=lambda v : v[1])

assigned = [0 for _ in range(n)]
available = deque([(1, 0)])
room_index = 1
for check in s:
    a, b, c = check
    if c == -1:
        if len(available) == 0:
            room_index += 1
            available.append((room_index, 0))
        else:
            room = available.pop()
            available.append(room)
            if room[1] == b:
                room_index += 1
                available.append((room_index, 0))
            
        assigned[a] = available.pop()[0]
    else:
        available.appendleft((assigned[a], b))

print(room_index)
print(*assigned)
